Add parser for unsaved Windows Notepad tabs

joost-j opened the pull request:
Just like Notepad++, the Windows Notepad application for Windows 11 is now able to restore unsaved tabs when you re-open the application. [This blog](https://medium.com/@mahmoudsoheem/new-digital-forensics-artifact-from-windows-notepad-527645906b7b) explains where it is stored and that you should be able to somehow view the contents. 

It turned out that there was a bit more to it than running `strings` or `grep` on it. The application stores the tabs in different formats, depending on the size and some other unknown factors. I've even encountered a file where 26 characters of text were encoded as 34(?!)  separate blocks; each block containing a length field, a single character and a CRC32 checksum of that very small block. Using `grep` or `strings` on that file would not have yielded any results. Information stored in these Notepad tabs may be helpful during forensic investigations and/or incident response cases.

The file format uses LEB128 variable-integer encoding for the block sizes, which is not yet present in the `dissect` framework. Therefore, this PR depends on https://github.com/fox-it/dissect.cstruct/pull/69, so in the end it depends on a new major release of `dissect.cstruct`.

The `dissect/target/plugins/aps/texteditor` folder was created, with corresponding new record types, so unsaved tabs from other text editors can also be added in the future.

Horofic commented on tests/plugins/apps/texteditor/test_texteditor.py:
```diff
@@ -0,0 +1,61 @@
+        "ba291ccd-f1c3-4ca8-949c-c01f6633789d.bin": (text5 * 5),
+        "e609218e-94f2-45fa-84e2-f29df2190b26.bin": (text6 * 1260),
+        "3d0cc86e-dfc9-4f16-b74a-918c2c24188c.bin": loremipsum,
+        "wrong-checksum.bin": "",  # only added to check for corrupt checksum, not validity
+        "cfe38135-9dca-4480-944f-d5ea0e1e589f.bin": (loremipsum * 37)[:-2],  # removed the two newlines in this file
+    }
+
+    tabcache = absolute_path("_data/plugins/apps/texteditor/windowsnotepad/")
+
+    user = target_win_users.user_details.find(username="John")
+    tab_dir = user.home_path.joinpath(windowsnotepad.WindowsNotepadPlugin.DIRECTORY)
+
+    fs_win.map_dir("Users\\John", tmp_path)
+
+    for file in file_text_map.keys():
+        tab_file = str(tab_dir.joinpath(file))[3:]
+        fs_win.map_file(tab_file, os.path.join(tabcache, file))
+
+    target_win.add_plugin(windowsnotepad.WindowsNotepadPlugin)
+
+    records = list(target_win.windowsnotepad.tabs())
+
+    # Check the amount of files
+    assert len(list(tab_dir.iterdir())) == len(file_text_map.keys())
+
+    # Only one should not be parsed correctly, without errors/warnings
+    assert len(records) == len(file_text_map.keys()) - 1
+
+    # One file should not return any contents, there should be an entry for this in the logging.
+    assert "CRC32 checksum mismatch in file: wrong-checksum.bin" in caplog.text
+    assert (
+        "CRCMismatchException: CRC32 mismatch in single-block file. expected=deadbeef, actual=a48d30a6" in caplog.text
+    )
+
+    # The recovered content in the records should match the original data, as well as the length
+    for rec in records:
+        assert rec.content == file_text_map[rec.filename]
+        assert len(rec.content) == len(file_text_map[rec.filename])
```
```suggestion
        "wrong-checksum.bin": text4,  # only added to check for corrupt checksum, not validity
        "cfe38135-9dca-4480-944f-d5ea0e1e589f.bin": (loremipsum * 37)[:-2],  # removed the two newlines in this file
    }

    tabcache = absolute_path("_data/plugins/apps/texteditor/windowsnotepad/")

    user = target_win_users.user_details.find(username="John")
    tab_dir = user.home_path.joinpath(
        "AppData/Local/Packages/Microsoft.WindowsNotepad_8wekyb3d8bbwe/LocalState/TabState"
    )

    fs_win.map_dir("Users\\John", tmp_path)

    for file in file_text_map.keys():
        tab_file = str(tab_dir.joinpath(file))[3:]
        fs_win.map_file(tab_file, os.path.join(tabcache, file))

    target_win.add_plugin(windowsnotepad.WindowsNotepadPlugin)

    records = list(target_win.windowsnotepad.tabs())

    # Check the amount of files
    assert len(list(tab_dir.iterdir())) == len(file_text_map.keys())
    assert len(records) == len(file_text_map.keys())

    # One file should still return contents, but there should be an entry for in the logging for a CRC missmatch.
    assert "CRC32 mismatch in single-block file: wrong-checksum.bin expected=deadbeef, actual=a48d30a6" in caplog.text

    # The recovered content in the records should match the original data, as well as the length
    for rec in records:
        assert rec.content == file_text_map[rec.path.name]
        assert len(rec.content) == len(file_text_map[rec.path.name])
```

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,174 @@
+        else:
+            # Reconstruct the text of the multi_block_entry variant
+            # CRC32 is calculated based on the entire header, up to the point where the CRC32 value is stored
+            assert tab.header_crc[0].crc32 == _calc_crc32(tab.dumps()[3 : tab.dumps().index(tab.header_crc[0].crc32)])
```
Refactored it into a warning log

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,174 @@
+
+        tab = c_windowstab.tab(fh)
+
+        if tab.len1 != 0:
```
Well, it does actually specify the length in some cases :smile: Except when it's nonzero - then the file consists of multiple blocks. Maybe `optional_total_length1` is a good name.

Schamper commented on dissect/target/exceptions.py:
```diff
@@ -114,3 +114,7 @@ class RegistryCorruptError(RegistryError):
     """An error occurred during configuration parsing."""
+
+
+class CRCMismatchException(Error):
+    """A mismatch between CRC checksums has occurred."""
```
I'd move this out of `dissect.target.exceptions` and into your plugin. Maybe even get rid of it and just raise a `PluginError` with an appropriate message.

joost-j replied:
This Exception is no longer used, so I'll just move it back to the shimcache plugin

Schamper commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,184 @@
+import zlib
+from typing import BinaryIO, Iterator
+
+from dissect import cstruct
```
```suggestion
from dissect.cstruct import cstruct
```

Schamper commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,184 @@
+        Returns:
+            A TextEditorTabRecord containing information that is in the tab.
+        """
+        fh: BinaryIO = file.open(mode="rb")
```
```suggestion
        fh: BinaryIO = file.open("rb")
```

It doesn't matter much in Dissect now, but can you use a context manager?

joost-j replied:
Yes, refactored it to use a context manager.

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,184 @@
+                                                           // Otherwise, it will parse the individual blocks.
+};
+"""
+TextEditorTabRecord = create_extended_descriptor([UserRecordDescriptorExtension])(
+    "texteditor/windowsnotepad/tab", GENERIC_TAB_CONTENTS_RECORD_FIELDS
+)
+
+c_windowstab = cstruct.cstruct()
+c_windowstab.load(c_def)
```
True, but all the data was stored in `AppData/Local/Packages/Microsoft.WindowsNotepad_` which means that it's the newer Windows Notepad App. It's different from the OG `notepad.exe` so that's why I made this distinction. Therefore I'd opt for just leaving it like this.

Schamper replied:
The rest of the suggestion still applies 😄

joost-j replied:
Ah right, now I see! :smile:

joost-j replied:
Changed it now :)

joost-j commented:
Refactored the code to work with cstruct v3 after https://github.com/fox-it/dissect.cstruct/pull/73 has been merged.

Schamper commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,182 @@
+
+def _calc_crc32(data: bytes) -> bytes:
+    """Perform a CRC32 checksum on the data and return it as bytes."""
+    return zlib.crc32(data).to_bytes(length=4, byteorder="big")
```
I assume the reason you return this as bytes is because for some reason the checksums are big endian? Maybe a comment in the structure would be nice if that's the case.

Nordgaren commented:
Heyo! You guys still working on this? I have gotten quite a bit labeled in my tabstate util. Found you guys mentioned this on John Hammonds tweet about the video going over this format. 

I know there is a crc32 after the weird timestamp, but I have not implemented that.

I was wondering, because looking at the code you guys have, so far, there might be a few things you guys are missing. I don't really understand how this package interprets the C code, but if it's just like normal C structures, I think the magic and header start might be off, [here](https://github.com/joost-j/dissect.target/blob/9df665ba2ea8f2360d0f95730b17258b1779f1c7/dissect/target/plugins/apps/texteditor/windowsnotepad.py#L41).

This format seems to change a lot, depending on the "state" it is in (new tab, unsaved tab and saved tab). I was wondering if maybe @joost-j might have some more work done on this, and maybe we can collaborate to fill out the rest? Here is my [tabstate util](https://github.com/Nordgaren/tabstate-util) for Rust.

joost-j commented:
Hi @Nordgaren, I think it's a great idea to collaborate on this! I have observed different storage formats as well, but could not yet link them to any action with regards to the "state" yet, so I'm really interested in that as well. Let's try to fill out the rest indeed!

As mentioned in your repo, I'll contact you on Discord.

daddycocoaman commented:
Hi,

I recently caught on to the Notepad tab hype and this plugin seems premature, IMO. It appears that Notepad tab files are a single format with three states:

- Unsaved file
- Saved file
- Saved file with unsaved data

The file extension can also slightly change depending on if the file is open somewhere else or not (`.tmp` files vs just `bin`). 

As it's written, this plugin will attempt to parse unsaved and saved files and will miss a lot of things, even in the unsaved state. For example, these structs can't be right at all:

```c
struct multi_block_entry {
    uint16    offset;
    uleb128   len;
    wchar     data[len];
    char      crc32[4];
};
struct single_block_entry {
    uint16    offset;
    uleb128   len;
    wchar     data[len];
    char      unk1;
    char      crc32[4];
};
```
In a large file, the offset cannot be a uint16 (65535 max size) because files can have more than 65535 characters. It's more than likely an `uleb128`. Additionally, these structs don't address what happens when a character is deleted, because the structure is different. If I add the letter `r` at offset 2 in the unsaved state, the file will append:  `02 00 01 72 00 A0 29 D5 3E`

Where 02 is the offset, 00 is "unknown", 01 is the uleb128 length, `72 00` is the wchar, then the crc32.

But if I remove that `r`, the file appends: `02 01 00 E5 DE 3C 3D`. Same offset "02" but the next value is a "01" which is the length of the item I removed (if you remove two character, it'll be two). What this all equates to is that if the second `uleb128` value is 00, then a character has been added. If it's anything other than zero, characters have been removed by that length.

With all that said, I've actually managed to figure out the notepad tab format (with the exception of a single u8 field) in all three states. I'll be posting about it soon and I'd recommend waiting before adding this plugin.

Schamper commented:
@daddycocoaman will you be contributing your findings?

Nordgaren commented:
> @daddycocoaman will you be contributing your findings?

I have given all of this data to Joost, myself, a while ago. I will be making a full write-up shortly. I didn't have time when I told him I would, but, everything listen here is in my util, except the crc. I just haven't taken the time to put it in yet. Unexpected life events. haha.

So you guys should be able to implement what I have so far, and he markdown file should follow this weekend. Maybe even a .bt file if I can manage!

Nordgaren commented:
> Hi,
> 
> I recently caught on to the Notepad tab hype and this plugin seems premature, IMO. It appears that Notepad tab files are a single format with three states:
> 
>     * Unsaved file
> 
>     * Saved file
> 
>     * Saved file with unsaved data
> 
> 
> The file extension can also slightly change depending on if the file is open somewhere else or not (`.tmp` files vs just `bin`).
> 
> As it's written, this plugin will attempt to parse unsaved and saved files and will miss a lot of things, even in the unsaved state. For example, these structs can't be right at all:
> 
> ```c
> struct multi_block_entry {
>     uint16    offset;
>     uleb128   len;
>     wchar     data[len];
>     char      crc32[4];
> };
> struct single_block_entry {
>     uint16    offset;
>     uleb128   len;
>     wchar     data[len];
>     char      unk1;
>     char      crc32[4];
> };
> ```
> 
> In a large file, the offset cannot be a uint16 (65535 max size) because files can have more than 65535 characters. It's more than likely an `uleb128`. Additionally, these structs don't address what happens when a character is deleted, because the structure is different. If I add the letter `r` at offset 2 in the unsaved state, the file will append: `02 00 01 72 00 A0 29 D5 3E`
> 
> Where 02 is the offset, 00 is "unknown" (not really unknown, read below), 01 is the uleb128 length, `72 00` is the wchar, then the crc32.
> 
> But if I remove that `r`, the file appends: `02 01 00 E5 DE 3C 3D`. Same offset "02" but the next value is a "01" which is the length of the item I removed (if you highlight and delete multiple characters, it'll be whatever the length of the deleted characters was).
> 
> What this all equates to is that if the second `uleb128` value is 00, then a character has been added. If it's anything other than zero, characters have been removed by that length.
> 
> With all that said, I've actually managed to figure out the notepad tab format (with the exception of a single u8 field) in all three states. I'll be posting about it soon and I'd recommend waiting before adding this plugin.

I can confirm most of what you said here. It is uleb128, as Joost identified.  You got the saved states, pretty much. There is also the unsaved new tab and like a "soft save", which happens when notepad closes, but the tab stays open without a filepath. This will write all of the buffer contents to the file, instead of the weird keystroke meme that it seems to be in the new tab state.

Did you get the entire metadata structure in the saved file state? That is pretty much the only data I haven't figured out all the way, but I know it's size.

you can checkout tabstate-util on my GitHub, if you want to cross reference your findings. Might be good. There's some weird curveballs in this format.

Nordgaren commented:
There is also [this issue](https://github.com/Nordgaren/tabstate-util/issues/1) on my repo which has a lot of good information

daddycocoaman commented:
> There is also [this issue](https://github.com/Nordgaren/tabstate-util/issues/1) on my repo which has a lot of good information

That is a great thread. To answer your question, yes I have all the fields of the saved state, including the cursor locations, timestamp, and a few other things in the format. There's no special  delimiters in the format. Everything has a specific value. 

To answer @Schamper, I hadn't actually heard of dissect before the comments on the John Hammond video. I think I might just put out the format and let people adopt it however they liked. Personally, as a red teamer, I have my own reasons. 😂

Nordgaren commented:
> > There is also [this issue](https://github.com/Nordgaren/tabstate-util/issues/1) on my repo which has a lot of good information
> 
> That is a great thread. To answer your question, yes I have all the fields of the saved state, including the cursor locations, timestamp, and a few other things in the format. There's no special delimiters in the format. Everything has a specific value.
> 
> To answer @Schamper, I hadn't actually heard of dissect before the comments on the John Hammond video. I think I might just put out the format and let people adopt it however they liked. Personally, as a red teamer, I have my own reasons. 😂

What did you find the single byte and then 4 byte int that mirrors it around the cursor start and end, to be?

Paradoxis commented:
@daddycocoaman 

> Personally, as a red teamer, I have my own reasons. 😂

Heya, fellow red teamer here: Could you maybe elaborate on this part? I'm genuinely curious why you wouldn't just share the information if you have it. 

Dissect isn't just a blue-team tool, I use it as a red teamer myself all the time as it ships with some pretty sick file parsing capabilities (especially nice if you're don't have 4TB of bandwidth to spare when all you need is like 200 bytes or something). Hell, I'd like to see this plugin get implemented as well, can think of some pretty cool stuff you could do with it

If it's losing out on credit you're worried about, external contributors are always credited (as I was [here](https://github.com/fox-it/dissect/releases/tag/3.9) and [here](https://github.com/fox-it/dissect/releases/tag/3.11)) :)

JustArion commented:
Hi there, ogmini and I both have some implementations of the tabs to near completion. ogimini has the C# implementation [here](https://github.com/ogmini/Notepad-Tabstate-Buffer?tab=readme-ov-file#overall-behavior) and I have the ImHex pattern implementation [here](https://github.com/JustArion/Notepad-Tabstate-Buffer?tab=readme-ov-file#file-tab). Both arn't 100% complete yet but **most** of it is.

Nordgaren commented:
My [`.bt`](https://github.com/Nordgaren/tabstate-util/blob/8e30c391e2e437ea57caf206f42ce5e777f5361c/TabState.bt#L1-L0) is uploaded, now.

It's missing new tab state files and tabstate files with extra buffers after the main one, for now, but it should be otherwise accurate.

ogmini commented:
> > Heya, fellow red teamer here: Could you maybe elaborate on this part? I'm genuinely curious why you wouldn't just share the information if you have it.
> 
> I'm putting out a blog post through my employer (hopefully reviewed and posted this week once I submit it today). It's less about credit on this repo and more like making sure the work I put in to RE the format is written up and distributed more formally.
> 
> > What did you find the single byte and then 4-byte int that mirrors it around the cursor start and end, to be?
> 
> Sorry, I misspoke when I said "all" (I was trying to be careful about not saying that). At any rate, the bytes after the cursor end are not a single field, but 4 separate byte fields that represent different boolean options. I've labeled them as `wordWrapEnabled`, `rightToLeftEnabled`, `showUnicodeControlChars`, and I'm missing the last one at the moment.
> 
> The byte before the cursor start always appear to be ``01` and is the field I mentioned earlier that I hadn't figured out. There are a couple of other boolean options related specifically to the Notepad UWF app but I have no idea how to configure them in either Notepad or the registry (like GhostFile or ClassicEditor, which I thought would mean using the old Notepad). I even tried turning off the integrated Copilot 🥲

Ah! More Notepad options... Hadn't thought to even test those yet. I hope you'll link the blog post when its published. This has been an interesting exercise and learning experience for me. 

Not to change topics, have you looked at the Windowstate files? I started to take a stab at those to stop from fixating on the Tabstate files too much. It stores window size and position as one would guess. I've only started trying to figure out the rest of the file. https://github.com/ogmini/Notepad-Windowstate-Buffer

Nordgaren commented:
> > Heya, fellow red teamer here: Could you maybe elaborate on this part? I'm genuinely curious why you wouldn't just share the information if you have it.
> 
> I'm putting out a blog post through my employer (hopefully reviewed and posted this week once I submit it today). It's less about credit on this repo and more like making sure the work I put in to RE the format is written up and distributed more formally.
> 
> > What did you find the single byte and then 4-byte int that mirrors it around the cursor start and end, to be?
> 
> Sorry, I misspoke when I said "all" (I was trying to be careful about not saying that). At any rate, the bytes after the cursor end are not a single field, but 4 separate byte fields that represent different boolean options. I've labeled them as `wordWrapEnabled`, `rightToLeftEnabled`, `showUnicodeControlChars`, and I'm missing the last one at the moment.
> 
> The byte before the cursor start always appear to be ``01` and is the field I mentioned earlier that I hadn't figured out. There are a couple of other boolean options related specifically to the Notepad UWF app but I have no idea how to configure them in either Notepad or the registry (like GhostFile or ClassicEditor, which I thought would mean using the old Notepad). I even tried turning off the integrated Copilot 🥲

Amazing. Thank you! So now we just need the byte before the cursors and the byte at the end of those bools, then! Thank you for sharing! I will go label the 3 bools. The last byte could also be padding, maybe? Also, did you find anything for the `0x00` that comes after the sha256 hash and before the unknown `0x01` before the cursors?

daddycocoaman commented:
I need to reconfirm but I think that first one is a null terminator for the SHA256. I remember seeing a comparison for 0x00 when it was comparing hashes but that could have been the data. 

Also, unfortunately, it's not the only thing left. The third byte in the format after "NP" is not part of the magic header. It's used by the .0 and .1 temp state files for unsaved files.

ogmini commented:
> Also, unfortunately, it's not the only thing left. The third byte in the format after "NP" is not part of the magic header. It's used by the .0 and .1 temp state files for unsaved files.

I've tagged that as the sequence number for the .0 and .1 files. It goes up incrementally. It is also a uLEB128.

https://github.com/ogmini/Notepad-Tabstate-Buffer?tab=readme-ov-file#0bin--1bin

Nordgaren commented:
> > Also, unfortunately, it's not the only thing left. The third byte in the format after "NP" is not part of the magic header. It's used by the .0 and .1 temp state files for unsaved files.
> 
> I've tagged that as the sequence number for the .0 and .1 files. It goes up incrementally. It is also a uLEB128.
> 
> https://github.com/ogmini/Notepad-Tabstate-Buffer?tab=readme-ov-file#0bin--1bin
> 
> *Edit
> 
> Oh, I think I misread your post. You already know about the .0 and .1 file. I've still assumed it to be a sequence number for the bin file. Just that it always appears to be 0x00.

I have actually noticed that the the 4th byte in the file, which is supposed to be the saved state, is also the count of remaining characters in a file, I think? Could be a uleb128, too, as you mentioned. the 3rd byte in the file does seem to change, and I think I have seen it change with the correct magic, as well. I can't remember for certain. Will try to dig a bit, shortly!

Nordgaren commented:
> I need to reconfirm but I think that first one is a null terminator for the SHA256. I remember seeing a comparison for 0x00 when it was comparing hashes but that could have been the data.
> 
> Also, unfortunately, it's not the only thing left. The third byte in the format after "NP" is not part of the magic header. It's used by the .0 and .1 temp state files for unsaved files.

It would be weird for it to be a null terminator for the sha 256 hash, though, as it is a fixed sized hash in the file. 32 bytes, or 256 bits. But I also wouldn't put it past microsoft. It could also be padding, as well? One thing I find weird is that they are using a varint for a fixed sized int. I assume that is for future proofing, though. So maybe they just have a fixed sized varint (lol) and some extra padding to compensate? That also seems like an out there idea, but just throwing it out there in case it helps anyone else see something different.

So far it doesn't look like they care about padding or even alignment, tbh.

joost-j commented on dissect/target/plugins/apps/texteditor/texteditor.py:
```diff
@@ -0,0 +1,17 @@
+
+GENERIC_TAB_CONTENTS_RECORD_FIELDS = [
+    ("string", "content"),
+    ("string", "content_length"),
```
No, not really. I'll leave it out. I initially figured it would be usefil in e.g. `target-query ... | rdump -s "r.content_length >= 200"` to look for long texts, but you can just do `"len(r.content) >= 2--"` as well..

joost-j commented:
Fixed the remaining code comments as of now @Schamper @Horofic. However, will include more functionality later this week, as a lot more about this file format is now known.

joost-j commented:
Added some more test cases, including some with the application closed/opened and including some with character deletions. Also simplified the code a bit; refactored some unnecessary complex operations and reduced the amount of structs.

The main goal for this plugin is to recover the text contents of the tab files and be able to parse and collect those during IR / Forensic investigations. So there are still some `unk` variables in the struct, of which I don't know all the details, but for this use-case that seems to be acceptable. Again, thanks all for the suggestions!

Horofic commented:
@daddycocoaman do you have a possible ETA when your research will be published?

ogmini commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,213 @@
+    uleb128     fileSize;
+    uleb128     encoding;
+    uleb128     carriageReturnType;
+    uleb128     timestamp; // Windows Filetime format (not unix timestamp)
```
From my and @JustArion testing, the timestamp is a uleb128. If we assume a uint64, it doesn't provide a valid Win32 filetime.  

![image](https://github.com/fox-it/dissect.target/assets/112595633/ef622824-192a-434e-a3ba-d8fa32322d74)

In the above example, we have A2 AC 8A 9E D3 A1 A3 ED 01 or 133574226803332642 in uleb128. This converts to 4/12/2024 7:11:20 PM in GMT. https://www.epochconverter.com/ldap 

This matches with the modified time on the original file. Which is 4/12/2024 3:11:20 PM in EST

![image](https://github.com/fox-it/dissect.target/assets/112595633/4ed77abe-d008-41d3-a5c9-574d193886ea)

This also lines up with the Filetime structure in the Win32 API (https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-filetime) which specifies a 64-bit value.

Horofic replied:
Thanks for the clarification @ogmini <3! Im keeping the comment unresolved for now, because I believe this is perfect info that could be returned to the analist.

joost-j replied:
Good suggestion! In the latest commit, I added a separate record that also presents the `saved_path`, `sha256` and `modification_time` fields to the analyst.

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,213 @@
+                        for idx in range(data_entry.nAdded):
+                            text.insert(data_entry.offset + idx, data_entry.data[idx])
+
+                    elif data_entry.nDeleted > 0:
```
Fixed, added a `--include-deleted-content` flag to the plugin

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,213 @@
+
+                # First, parse 4 unknown bytes. These likely
+                # hold some addition information about the tab (view options etc.)
+                unknown_bytes = fh.read(4)
```
Yes, moved all the `fh.read()` back into the structures

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,213 @@
+                        break
+
+                    # Each block has a CRC32 value in big-endian appended to the block
+                    defined_crc32 = fh.read(4)
```
Yes, I added those reads back into the structs

daddycocoaman commented:
> @daddycocoaman do you have a possible ETA when your research will be published?

Sorry for the delay. At some point during this, I think the tab state format added a few new fields and it broke my ImHex parser. For example, there is now one for spell check being enabled that comes right before the length of the content. I'm looking into the new fields in my spare time.

ogmini commented:
> > @daddycocoaman do you have a possible ETA when your research will be published?
> 
> Sorry for the delay. At some point during this, I think the tab state format added a few new fields and it broke my ImHex parser. For example, there is now one for spell check being enabled that comes right before the length of the content. I'm looking into the new fields in my spare time.

What version of Notepad is it reporting?

ogmini commented:
> > > @daddycocoaman do you have a possible ETA when your research will be published?
> > 
> > 
> > Sorry for the delay. At some point during this, I think the tab state format added a few new fields and it broke my ImHex parser. For example, there is now one for spell check being enabled that comes right before the length of the content. I'm looking into the new fields in my spare time.
> 
> What version of Notepad is it reporting?

Had a few moments to quickly poke around. There definitely are some changes to the format that appears to lineup with new options in what I call the options block. My test machine is running the stable build of Windows 11 23H2 OS Build - 22631.3527 and Windows Notepad version 11.2402.22.0. The format has changed but I don't have the options to enable/disable spellcheck or autocorrect within Windows Notepad. I'll spin up another test system later on a Release Preview build to see if I get the options.  

https://blogs.windows.com/windows-insider/2024/03/21/spellcheck-in-notepad-begins-rolling-out-to-windows-insiders/

joost-j commented:
Looking at your screenshots @ogmini, the byte at offset `0A` is different between the versions and may indicate some kind of `optionsVersion` number. Thank you both again for the updates, I'll edit my dissect parser accordingly as well!

ogmini commented:
> Looking at your screenshots @ogmini, the byte at offset `0A` is different between the versions and may indicate some kind of `optionsVersion` number. Thank you both again for the updates, I'll edit my dissect parser accordingly as well!

That was one of my theories. But it doesn't make total sense. Why would you have the version so late in the stream? The files do store content length, so it knows how far to read the stream. Is it possible that the byte at offset `0A` is letting us know how many more bytes to read? We'll probably need to wait for MS to make more changes in the future to really verify.

Consolidated screenshot with some coloring for easier reading.
 
![image](https://github.com/fox-it/dissect.target/assets/112595633/95ae6e98-639f-4179-a354-6cb998c4e648)

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,213 @@
+    uleb128     encoding;
+    uleb128     carriageReturnType;
+    uleb128     timestamp; // Windows Filetime format (not unix timestamp)
+    char        sha256[32];
```
Yes, added the sha256 hash to the record as well.

joost-j commented:
I added the `optionsVersion` field in the latest commit. Although we might not be 100% sure if this is correct, we can always fill in these gaps later on. At least, the dissect parser now seems to be able to parse the newer format as well. The parser is now also able to handle variable-length data buffers that is passed after a fixed-length data buffer.

joost-j commented:
Although the file format is still changing and we might not have uncovered 100% of the file format, I feel this is a great example of online collaboration to get things done. I therefore submitted a proposal for a talk at the SANS summit in Prague in September 2024, which got accepted last week. :tada:

The goal of the talk is not to explain the file format in full detail, nor to take full credit of uncovering the file format, but to encourage and inspire people to take a (deep) dive into unknown file formats and share the findings with the entire DFIR community! Of course I'll be sure to mention your contributions @Nordgaren @daddycocoaman @ogmini @JustArion, as well as John Hammond's video. And of course, if one of you happens to be there in person, let's discuss this irl!

ogmini commented:
> Although the file format is still changing and we might not have uncovered 100% of the file format, I feel this is a great example of online collaboration to get things done. I therefore submitted a proposal for a talk at the SANS summit in Prague in September 2024, which got accepted last week. 🎉
> 
> The goal of the talk is not to explain the file format in full detail, nor to take full credit of uncovering the file format, but to encourage and inspire people to take a (deep) dive into unknown file formats and share the findings with the entire DFIR community! Of course I'll be sure to mention your contributions @Nordgaren @daddycocoaman @ogmini @JustArion, as well as John Hammond's video. And of course, if one of you happens to be there in person, let's discuss this irl!

Very cool, I hope you can share your slides or talk after the summit. I'm also going to be submitting a talk to BSides-NYC mainly talking about my experience, what I've learned, and my process. Hopefully it is accepted. I'll also make sure to mention you all.  

I haven't had much of a chance to update this thread since this isn't remotely my day job. I've been working on trying to automate generation of test state files with known parameters to more easily detect changes from future updates.  I've also been poking at the application hive which is stored as the `settings.dat` file in the Settings folder. Nothing particularly mind-blowing in that file. 

I continue to update my notes here: https://github.com/ogmini/Notepad-State-Library

Horofic commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,377 @@
+                content += data_entry.data
+
+                # The hasRemainingVariableDataBlocks indicates whether more data will follow after this single block
+                if data_entry.hasRemainingVariableDataBlocks == 1:
```
Cant you just use this value to test for remaining data? Instead of using the `has_remaining_data` variable.

Horofic commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,377 @@
+    return zlib.crc32(data).to_bytes(length=4, byteorder="big")
+
+
+class WindowsNotepadTabContent:
```
It's a better idea to return a `WindowsNotepadTab` object, and leaving the creation of the records to the plugin code. This way, when I use this class to parse a Windows Notepad tab I have access to the low-level attributes like `tab. file_header.updateNumber` in my own code instead of getting just a `Record` representation of the tab.

Horofic commented:
@daddycocoaman did you make any progress in the meantime or a estimation on a possible release? I'm eager on how we can improve the implementation further!

joost-j commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,377 @@
+
+                # Add the deleted content, if specified
+                if include_deleted_content:
+                    text += " --- DELETED-CONTENT: "
```
Good point, I'll add an extra record field in the output

Schamper commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,341 @@
+        if not self.users_tabs:
+            raise UnsupportedPluginError("No Windows Notepad tab files found")
+
+    @export(record=DynamicDescriptor(["path", "datetime", "string"]))
```
```suggestion
    @export(record=[WindowsNotepadSavedTabRecord, WindowsNotepadUnsavedTabRecord]))
```

Schamper commented on dissect/target/plugins/apps/texteditor/windowsnotepad.py:
```diff
@@ -0,0 +1,341 @@
+    GENERIC_TAB_CONTENTS_RECORD_FIELDS + WINDOWS_SAVED_TABS_EXTRA_FIELDS,
+)
+
+c_windowstab = cstruct()
+c_windowstab.load(c_def)
```
```suggestion
c_windowstab = cstruct().load(windowstab_def)
```
